File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/combinationsum/CombinationSum.java
package combinationsum;

import java.util.*;

public class CombinationSum {

    private final int[] candidates;
    private final int target;
    private final List<List<Integer>> solutions = new ArrayList<>();

    public CombinationSum(int[] candidates, int target) {
        if (candidates == null || candidates.length == 0)
            throw new IllegalArgumentException("Candidates must not be empty");
        this.candidates = candidates.clone();
        Arrays.sort(this.candidates);
        this.target = target;
    }

    public void solveAndPrint() {
        backtrack(new ArrayList<>(), 0, 0);
        System.out.printf("Found %d combination(s) summing to %d%n%n", solutions.size(), target);

        int limit = Math.min(solutions.size(), 8);
        for (int i = 0; i < limit; i++) {
            List<Integer> sol = solutions.get(i);
            System.out.printf("#%d: %s = %d%n", i + 1, sol, sol.stream().mapToInt(x -> x).sum());
        }
        if (solutions.size() > limit) {
            System.out.printf("... and %d more.%n", solutions.size() - limit);
        }
    }

    private void backtrack(List<Integer> current, int start, int sum) {
        if (sum == target) {
            solutions.add(new ArrayList<>(current));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            if (sum + candidates[i] > target) break;
            current.add(candidates[i]);
            backtrack(current, i, sum + candidates[i]);
            current.remove(current.size() - 1);
        }
    }
}

File: 29Items-Projects/7Algorithms/6-Backtracking/python/combination_sum.py
"""Combination Sum — find all unique combinations that sum to target (unlimited reuse)."""


class CombinationSum:
    def __init__(self, candidates: list[int], target: int):
        if not candidates:
            raise ValueError("Candidates must not be empty")
        self.candidates = sorted(candidates)
        self.target = target
        self.solutions: list[list[int]] = []

    def solve_and_print(self) -> None:
        self._backtrack([], 0, 0)
        print(f"Found {len(self.solutions)} combination(s) summing to {self.target}\n")

        limit = min(len(self.solutions), 8)
        for i, sol in enumerate(self.solutions[:limit], 1):
            print(f"#{i}: {sol} = {sum(sol)}")
        if len(self.solutions) > limit:
            print(f"... and {len(self.solutions) - limit} more.")

    def _backtrack(self, current: list[int], start: int, total: int) -> None:
        if total == self.target:
            self.solutions.append(current[:])
            return
        for i in range(start, len(self.candidates)):
            if total + self.candidates[i] > self.target:
                break
            current.append(self.candidates[i])
            self._backtrack(current, i, total + self.candidates[i])
            current.pop()
